src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/RSAKeyPemTests.cs

```diff
@@ -0,0 +1,384 @@
+        {
+            string pem = @"
+-----BEGIN RSA PUBLIC KEY-----
+MEgCQQC3P1n17ovVXiS3/wKa0WqFQ8ltJT5UMZuTUyxBw8FHe4nbLS8z17modFhI
+4GqOaDtQRFEeG8o2JSfhfPQrOAYVAgMBAAE=
+-----END RSA PUBLIC KEY-----";
+
+            using RSA rsa = RSAFactory.Create();
+            rsa.ImportFromPem(pem);
+            RSAParameters rsaParameters = rsa.ExportParameters(false);
+
+            ImportExport.AssertKeyEquals(TestData.DiminishedDPParameters.ToPublic(), rsaParameters);
+        }
+
+        [Fact]
+        public static void ImportFromPem_RSAPrivateKey_PrecedingUnrelatedPem()
+        {
+            string pem = @"
+-----BEGIN CERTIFICATE-----
+MIICTzCCAgmgAwIBAgIJAMQtYhFJ0+5jMA0GCSqGSIb3DQEBBQUAMIGSMQswCQYD
+VQQGEwJVUzETMBEGA1UECAwKV2FzaGluZ3RvbjEQMA4GA1UEBwwHUmVkbW9uZDEY
+MBYGA1UECgwPTWljcm9zb2Z0IENvcnAuMSAwHgYDVQQLDBcuTkVUIEZyYW1ld29y
+ayAoQ29yZUZ4KTEgMB4GA1UEAwwXUlNBIDM4NC1iaXQgQ2VydGlmaWNhdGUwHhcN
+MTYwMzAyMTY1OTA0WhcNMTYwNDAxMTY1OTA0WjCBkjELMAkGA1UEBhMCVVMxEzAR
+BgNVBAgMCldhc2hpbmd0b24xEDAOBgNVBAcMB1JlZG1vbmQxGDAWBgNVBAoMD01p
+Y3Jvc29mdCBDb3JwLjEgMB4GA1UECwwXLk5FVCBGcmFtZXdvcmsgKENvcmVGeCkx
+IDAeBgNVBAMMF1JTQSAzODQtYml0IENlcnRpZmljYXRlMEwwDQYJKoZIhvcNAQEB
+BQADOwAwOAIxANrMIthuZxV1Ay4x8gbc/BksZeLVEInlES0JbyiCr9tbeM22Vy/S
+9h2zkEciMuPZ9QIDAQABo1AwTjAdBgNVHQ4EFgQU5FG2Fmi86hJOCf4KnjaxOGWV
+dRUwHwYDVR0jBBgwFoAU5FG2Fmi86hJOCf4KnjaxOGWVdRUwDAYDVR0TBAUwAwEB
+/zANBgkqhkiG9w0BAQUFAAMxAEzDg/u8TlApCnE8qxhcbTXk2MbX+2n5PCn+MVrW
+wggvPj3b2WMXsVWiPr4S1Y/nBA==
+-----END CERTIFICATE-----
+-----BEGIN RSA PRIVATE KEY-----
+MIIBOwIBAAJBALc/WfXui9VeJLf/AprRaoVDyW0lPlQxm5NTLEHDwUd7idstLzPX
+uah0WEjgao5oO1BEUR4byjYlJ+F89Cs4BhUCAwEAAQJBAK/m8jYvnK9exaSR+DAh
+Ij12ip5pB+HOFOdhCbS/coNoIowa6WJGrd3Np1m9BBhouWloF8UB6Iu8/e/wAg+F
+9ykCIQDzcnsehnYgVZTTxzoCJ01PGpgESilRyFzNEsb8V60ZewIhAMCyOujqUqn7
+Q079SlHzXuvocqIdt4IM1EmIlrlU9GGvAh8Ijv3FFPUSLfANgfOIH9mX7ldpzzGk
+rmaUzxQvyuVLAiEArCTM8dSbopUADWnD4jArhU50UhWAIaM6ZrKqC8k0RKsCIQDC
+yZWUxoxAdjfrBGsx+U6BHM0Myqqe7fY7hjWzj4aBCw==
+-----END RSA PRIVATE KEY-----";
+            using RSA rsa = RSAFactory.Create();
+            rsa.ImportFromPem(pem);
+            RSAParameters rsaParameters = rsa.ExportParameters(true);
+
+            ImportExport.AssertKeyEquals(TestData.DiminishedDPParameters, rsaParameters);
+        }
+
+        [Fact]
+        public static void ImportFromPem_RSAPrivateKey_PrecedingMalformedPem()
+        {
+            string pem = @"
+-----BEGIN CERTIFICATE-----
+$$ I AM NOT A PEM
+-----END CERTIFICATE-----
+-----BEGIN RSA PRIVATE KEY-----
+MIIBOwIBAAJBALc/WfXui9VeJLf/AprRaoVDyW0lPlQxm5NTLEHDwUd7idstLzPX
+uah0WEjgao5oO1BEUR4byjYlJ+F89Cs4BhUCAwEAAQJBAK/m8jYvnK9exaSR+DAh
+Ij12ip5pB+HOFOdhCbS/coNoIowa6WJGrd3Np1m9BBhouWloF8UB6Iu8/e/wAg+F
+9ykCIQDzcnsehnYgVZTTxzoCJ01PGpgESilRyFzNEsb8V60ZewIhAMCyOujqUqn7
+Q079SlHzXuvocqIdt4IM1EmIlrlU9GGvAh8Ijv3FFPUSLfANgfOIH9mX7ldpzzGk
+rmaUzxQvyuVLAiEArCTM8dSbopUADWnD4jArhU50UhWAIaM6ZrKqC8k0RKsCIQDC
+yZWUxoxAdjfrBGsx+U6BHM0Myqqe7fY7hjWzj4aBCw==
+-----END RSA PRIVATE KEY-----";
```

Comment:
Is it ambiguous on an RSA key for `EC PRIVATE KEY` and `RSA PRIVATE KEY`? What about `EC PRIVATE KEY` and `PRIVATE KEY` (whether or not the PKCS#8 has the RSA Oid)?

I'm not sure I actually care what the answer is, but I'd like to see a "not contextually applicable" format test that used a known-as-a-key-type format.

Reply:
We touched a bit on this in the original issue, the gist is:

1. For PKCS8/SPKI keys, we don't decode them for purposes of reporting a duplicate. If a PKCS1 and PKCS8 PEM are found, we assume dupe, even in the PKCS8 turns out to be a key for an entirely different algorithm.
2. If the labels are for two known-to-be-not-ambiguous keys, like EC PRIVATE KEY and RSA PRIVATE KEY, we won't report a duplicate. I think the rationale there was that we don't have a crystal ball for future algorithms. If we throw ambiguous for EC PRIVATE KEY for example while importing RSA, and eventually .NET gets enlightened with a new algorithm with PEM encoding (say ED25519 since it doesn't use the regular "EC" label), we wouldn't want to start breaking people's existing PEM aggregates that contained those algorithms we once ignored.

Reply:
Yep, those both sound reasonable (and what I felt the conclusions had been); I'm after the test that proves it for "known key types" vs "this algorithm's known key types" :smile:

Reply:
Regardless I will illustrate this better with tests.